### PR TITLE
Fix #4507 - fusion specific fields in export

### LINKED ISSUE #4507: Fusion specific fields missing from variantS export
It would be nifty to export things like exon number, orientation, maybe split read counts etc, and `fusion_genes` for fallback when they do not correspond to plain HGNC genes that we have loaded.

### PR DESCRIPTION
**Pull request by dnil**
This PR adds a functionality or fixes a bug.

- Fix #4507 - added fusion specific fields missing from variantS export.

Screenshot 2024-03-15 at 14 28 56

<details>
<summary>Testing on cg-vm1 server (Clinical Genomics Stockholm)</summary>

**Prepare for testing**
1. Make sure the PR is pushed and available on [Docker Hub](https://hub.docker.com/repository/docker/clinicalgenomics/scout-server-stage)
1. Fist book your testing time using the Pax software available at [https://pax.scilifelab.se/](https://pax.scilifelab.se). The resource you are going to call dibs on is `scout-stage` and the server is `cg-vm1`.
1. `ssh <USER.NAME>@cg-vm1.scilifelab.se`
1. `sudo -iu hiseq.clinical`
1. `ssh localhost`
1. (optional) Find out which scout branch is currently deployed on cg-vm1: `podman ps`
1. Stop the service with current deployed branch: `systemctl --user stop scout.target`
1. Start the scout service with the branch to test: `systemctl --user start scout@<this_branch>`
1. Make sure the branch is deployed: `systemctl --user status scout.target`
1. After testing is done, repeat procedure at [https://pax.scilifelab.se/](https://pax.scilifelab.se), which will release the allocated resource (`scout-stage`) to be used for testing by other users.
</details>

<details>
<summary>Testing on hasta server (Clinical Genomics Stockholm)</summary>

**Prepare for testing**
1. `ssh <USER.NAME>@hasta.scilifelab.se`
1. Book your testing time using the Pax software. `us; paxa -u <user> -s hasta -r scout-stage`. You can also use the WSGI Pax app available at [https://pax.scilifelab.se/](https://pax.scilifelab.se).
1. (optional) Find out which scout branch is currently deployed on cg-vm1: `conda activate S_scout; pip freeze | grep scout-browser`
1. Deploy the branch to test: `bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_scout -t scout -b <this_branch>`
1. Make sure the branch is deployed: `us; scout --version`
1. After testing is done, repeat the `paxa` procedure, which will release the allocated resource (`scout-stage`) to be used for testing by other users.
</details>

**How to test**:
1. how to test it, possibly with real cases/data

**Expected outcome**:
The functionality should be working
Take a screenshot and attach or copy/paste the output.

**Review:**
- [x] code approved by CR
- [x] tests executed by DN, CR

### CODE REVIEW

File: scout/server/blueprints/variants/controllers.py [hunk 1]
```diff
@@ -1024,21 +1026,17 @@ def generate(header, lines):
     )
 
 
-def download_variants(store, case_obj, variant_objs):
+def download_variants(
+    store: MongoAdapter, case_obj: dict, variant_objs: CursorType, category: Optional[str] = None
```
Comment (northwestwitch) [inline scout/server/blueprints/variants/controllers.py:1030]: 🥇

File: scout/server/blueprints/variants/controllers.py [hunk 2]
```diff
@@ -1024,21 +1026,17 @@ def generate(header, lines):
...
+) -> Response:
     """Download filtered variants for a case to an excel file
```
Comment (northwestwitch) [inline scout/server/blueprints/variants/controllers.py:1032]:
```suggestion
    """Download filtered variants for a case to a .csv file```

File: scout/server/blueprints/variants/controllers.py [hunk 3]
```diff
@@ -1060,84 +1058,172 @@ def generate(header, lines):
...
+        for transcript in gene.get("transcripts", []):
+            if "exon" in transcript:
+                if exon:
+                    exon = exon + " | "
+                exon = exon + transcript["exon"]
+
+    return exon
+
+
+def variant_export_lines_fusion(variant: dict, case_obj: dict) -> list:
```
Comment (northwestwitch) [inline scout/server/blueprints/variants/controllers.py:1130]:
```suggestion
def variant_export_lines_fusion(variant: dict, case_obj: dict) -> List[str]:
```

File: scout/server/blueprints/variants/controllers.py [hunk 4]
```diff
@@ -1060,84 +1058,172 @@ def generate(header, lines):
...
+        value = "N/A"
+    return value
+
+
+def get_fusion_exons(variant: dict) -> str:
+    """Get RNAfusion specific variant exons for all genes/transcripts listed. Join with pipe 'or' character."""
+    exon = ""
+    for gene in variant.get("genes", []):
+        for transcript in gene.get("transcripts", []):
+            if "exon" in transcript:
```
Comment (northwestwitch) [inline scout/server/blueprints/variants/controllers.py:1122]:
```suggestion
           if bool(transcript.get("exon")):
```

File: scout/server/blueprints/variants/controllers.py [hunk 5]
```diff
@@ -1060,84 +1058,172 @@ def generate(header, lines):
...
+                if exon:
```
Comment (northwestwitch) [inline scout/server/blueprints/variants/controllers.py:1123]:
```suggestion
```

File: scout/server/blueprints/variants/controllers.py [hunk 6]
```diff
@@ -1060,84 +1058,172 @@ def generate(header, lines):
...
+            if individual["individual_id"] != variant_gt["sample_id"]:
+                continue
+            variant_line.append(variant_gt.get("read_depth", "N/A"))
+            variant_line.append(variant_gt.get("split_read", "N/A"))
+            variant_line.append(variant_gt.get("ffpm", "N/A"))
+
+    return variant_line
+
+
+def variant_export_lines_cancer(variant: dict) -> list:
```
Comment (northwestwitch) [inline scout/server/blueprints/variants/controllers.py:1153]:
```suggestion
def variant_export_lines_cancer(variant: dict) -> List[str]:
```

File: scout/server/blueprints/variants/controllers.py [hunk 7]
```diff
@@ -1060,84 +1058,172 @@ def generate(header, lines):
...
+        variant_line.append(vaf_sample)
+
+    # ADD eventual COSMIC ID
+    cosmic_ids = variant.get("cosmic_ids") or ["-"]
+    variant_line.append(" | ".join(cosmic_ids))
+
+    return variant_line
+
+
+def variant_export_lines_rare(variant: dict, case_obj: dict) -> list:
```
Comment (northwestwitch) [inline scout/server/blueprints/variants/controllers.py:1180]:
```suggestion
def variant_export_lines_rare(variant: dict, case_obj: dict) -> List[str]:
```

File: scout/server/blueprints/variants/controllers.py [hunk 8]
```diff
@@ -1060,84 +1058,172 @@ def generate(header, lines):
...
+):
     """Get variants info to be exported to file, one list (line) per variant.
     Args:
         store(scout.adapter.MongoAdapter)
         case_obj(scout.models.Case)
         variants_query: a list of variant objects, each one is a dictionary
     Returns:
         export_variants: a list of strings. Each string  of the list corresponding to the fields
-                         of a variant to be exported to file, separated by comma
+                         of a variant to be exported to file, separated by comma.
```
Comment (northwestwitch) [inline scout/server/blueprints/variants/controllers.py:1213]: Remove Args and Returns and instead add return type to the function?

File: scout/server/blueprints/variants/controllers.py [hunk 9]
```diff
@@ -1223,7 +1309,7 @@ def variant_export_genes_info(store, gene_list, genome_build="37"):
     return gene_info
 
 
-def variants_export_header(case_obj):
+def variants_export_header(case_obj: dict, category: str = "snv") -> list:
```
Comment (northwestwitch) [inline scout/server/blueprints/variants/controllers.py:1312]:
```suggestion
def variants_export_header(case_obj: dict, category: str = "snv") -> List[str]:
```

**Review by northwestwitch — APPROVED** [review]
Great code refactoring! 
I've tested it on stage with several variant types and works just fine! 💯 

I've got very simple suggestions, type hints mostly

File: scout/server/blueprints/variants/controllers.py (hunk 3 above)
Comment (northwestwitch) [reply scout/server/blueprints/variants/controllers.py:1130]: Ah OK!

File: scout/server/blueprints/variants/controllers.py (hunk 2 above)
Comment (dnil) [reply scout/server/blueprints/variants/controllers.py:1032]: Yep, thank you!

File: scout/server/blueprints/variants/controllers.py (hunk 5 above)
Comment (dnil) [reply scout/server/blueprints/variants/controllers.py:1123]: Hm, no, that would mean all exon strings start with " | ". This is using `bool("") == False` to check if it is the first iteration of the loop.

File: scout/server/blueprints/variants/controllers.py (hunk 8 above)
Comment (dnil) [reply scout/server/blueprints/variants/controllers.py:1213]: And here we know it is a `List[str]`!

**Pull request merged**